lib/buffy/throttle_and_timed.ex

```diff
@@ -0,0 +1,387 @@
+      def init({key, args}) do
+        Process.send_after(self(), :timeout, unquote(throttle))
+        {:ok, %{key: key, args: args, work_status: :scheduled}}
+      end
+
+      @doc """
+      Function to invoke the throttle logic if process already exists.
+      It will invoke the throttle logic if `:work_status` in state is either `:complete` or `:scheduled_by_loop_interval`.
+      `:scheduled_by_loop_interval` is set by the empty inbox timeout, which should be overruled with any incoming message -
+      since the timeout is reset by the definition of empty inbox timeout, the throttle logic can safely be scheduled.
+
+      """
+      @impl GenServer
+      @spec handle_cast(:throttle, Buffy.ThrottleAndTimed.state()) :: {:noreply, Buffy.ThrottleAndTimed.state()}
+      def handle_cast(:throttle, %{work_status: :complete} = state) do
+        Process.send_after(self(), :timeout, unquote(throttle))
+        {:noreply, %{state | work_status: :scheduled}}
+      end
+
+      def handle_cast(:throttle, %{work_status: :scheduled_by_loop_interval} = state) do
+        Process.send_after(self(), :timeout, unquote(throttle))
+        {:noreply, %{state | work_status: :scheduled}}
+      end
+
+      def handle_cast(:throttle, %{work_status: _} = state) do
+        {:noreply, state}
+      end
+
+      @doc false
+      @impl GenServer
+      @spec handle_info(:timeout, Buffy.ThrottleAndTimed.state()) ::
+              {:noreply, Buffy.ThrottleAndTimed.state(), {:continue, :do_work}}
+      def handle_info(:timeout, %{key: key, args: args} = state) do
+        new_state = %{state | work_status: :in_progress}
+        {:noreply, new_state, {:continue, :do_work}}
+      end
+
+      @doc false
+      @impl GenServer
+      @spec handle_continue(do_work :: atom(), Buffy.ThrottleAndTimed.state()) ::
+              {:noreply, Buffy.ThrottleAndTimed.state()} | {:noreply, Buffy.ThrottleAndTimed.state(), timeout()}
+      def handle_continue(:do_work, %{key: key, args: args} = state) do
+        :telemetry.span(
+          [:buffy, :throttle, :handle],
+          %{args: args, key: key, module: __MODULE__},
+          fn ->
+            result = handle_throttle(args)
+            {result, %{args: args, key: key, module: __MODULE__, result: result}}
+          end
+        )
+
+        new_state = %{state | work_status: :complete}
+        maybe_add_inbox_timeout_and_update_work_status({:noreply, new_state})
+      rescue
+        e ->
+          Logger.error("Error in throttle: #{inspect(e)}")
+          new_state = %{state | work_status: :complete}
+          maybe_add_inbox_timeout_and_update_work_status({:noreply, new_state})
+      end
+
+      defp maybe_add_inbox_timeout_and_update_work_status(
+             {return_signal, %{work_status: work_status} = state} = return_tuple
+           ) do
+        loop_interval = unquote(loop_interval)
+
+        ThrottleAndTimed.maybe_add_inbox_timeout_and_update_work_status(loop_interval, return_tuple)
```

Comment:
```suggestion
        maybe_add_inbox_timeout_and_update_work_status(loop_interval, return_tuple)
```

Reply:
Ah this is somewhat my fault to name the functions the same - but `ThrottleAndTimed.maybe_add_inbox_timeout_and_update_work_status` points to the non-macro function that lives in the module.

Reply:
Function removed in ff1de5e